src/CodeGen_D3D12Compute_Dev.cpp

```diff
@@ -0,0 +1,1154 @@
+            {
+                // must pack/unpack elements to/from shared memory...
+                elements /= 2;
+                bytesize /= 2;
+                packing_factor *= 2;
+            }
+            stream << " [" << elements << "];\n";
+            // smallest possible pack type is a byte (no nibbles)
+            internal_assert( packing_factor <= 4 );
+        }
+        else
+        {
+            // fill-in __GROUPSHARED_SIZE_IN_BYTES later on when D3DCompile() is
+            // invoked in halide_d3d12compute_run(); must divide by 4 since 
+            stream << " [__GROUPSHARED_SIZE_IN_BYTES / 4];\n";
+        }
+        Allocation alloc;
+        alloc.type = op->type;
+        allocations.push(op->name, alloc);
+    }
+
+    // Emit the kernel function prototype
+
+    // Figure out the thread group size by traversing the stmt:
+    class FindThreadGroupSize : public IRVisitor
+    {
+        using IRVisitor::visit;
+        void visit(const For* loop)
+        {
+            if (!is_gpu_var(loop->name))
+                return loop->body.accept(this);
+            if (loop->for_type != ForType::GPUThread)
+                return loop->body.accept(this);
+            internal_assert(is_zero(loop->min));
+            int index = thread_loop_workgroup_index(loop->name);
+            user_assert(index >= 0) << "Invalid 'numthreads' index for loop variable '" << loop->name << "'.\n";
+            // if 'numthreads' for a given dimension can't be determined at code
+            // generation time, emit code such that it can be patched some point
+            // later when calling D3DCompile() / halide_d3d12compute_run()
+            numthreads[index] = 0;  // <-- 0 indicates 'undetermined'
+            const IntImm* int_limit = loop->extent.as<IntImm>();
+            if (nullptr != int_limit)
+            {
+                numthreads[index] = int_limit->value;
+                user_assert(numthreads[index] > 0) << "For D3D12Compute, 'numthreads[" << index << "]' values must be greater than zero.\n";
+            }
+            debug(4) << "Thread group size for index " << index << " is " << numthreads[index] << "\n";
+            loop->body.accept(this);
+        }
+        int thread_loop_workgroup_index(const string &name)
+        {
+            string ids[] = { ".__thread_id_x",
+                             ".__thread_id_y",
+                             ".__thread_id_z",
+                             ".__thread_id_w" };
+            for (auto& id : ids)
+            {
+                if (ends_with(name, id))
+                {
+                    return (&id - ids);
+                }
+            }
+            return -1;
+        }
+        friend class CodeGen_D3D12Compute_Dev::CodeGen_D3D12Compute_C;
+        int numthreads [3] = { 1, 1, 1 };
+    };
+    FindThreadGroupSize ftg;
+    s.accept(&ftg);
+    // for undetermined 'numthreads' dimensions, insert placeholders to the code
```

Comment:
Is there any reason not to always do this, instead of trying to figure out the threadgroup size statically?  Why not always just use these preprocessor directives and patch it in at shader compile time

Reply:
No real reason.
Originally, the cases I was testing all had `numthreads` statically determined at code-gen time.
Later on, while fixing the back-end to pass all correctness tests, I came across these cases where `numthreads` would only be known close to kernel dispatch time. 
I think there's no harm in having both, since having the value of numthreads stamped directly into the shader code can be useful for debugging.
It is also an interesting example of the IR visitor that I can refer to in the future.